File: app/src/main/java/com/examples/android/androidtrainingbzu/HR_MainActivity.java
package com.examples.android.androidtrainingbzu;

import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;

import com.examples.android.androidtrainingbzu.Models.Employee;
import com.examples.android.androidtrainingbzu.Utils.DataSource;
import com.examples.android.androidtrainingbzu.Utils.Utils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.LinkedList;

public class HR_MainActivity extends AppCompatActivity{
    public static LinkedList<Employee> employeeLinkedList = new LinkedList();
    private final String TAG = MainActivity.class.getSimpleName();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_hr__main);
    }

    public void startActivity() {
        Intent intent = new Intent(getApplicationContext(), EmployeeListActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }
    public void callInitialize(View view) {
        switch (view.getId()) {
            case R.id.btn_inialize:
                new AsyncInitializerTask().execute(DataSource.INITIALIZE);
//                employeeLinkedList.clear();
//                Employee e;
//                int empID = 1;
//                String[] depts = {"Accounting", "HR", "IT"};
//                int[] avatars = {R.drawable.emp_pic1, R.drawable.emp_pic2, R.drawable.emp_pic3, R.drawable.emp_pic4, R.drawable.emp_pic5, R.drawable.emp_pic6, R.drawable.emp_pic7};
//                for (int i = 0; i < 20; i++) {
//                    e = new Employee("Emplyee " + empID + " Name ", "Emplyee " + empID + " Email ", "Emplyee " + empID + " Phone ",
//                            "Emplyee " + empID + " Address ", depts[i % 3], empID++, avatars[i % 7], Utils.getDateFromString("22.05.2016"));
//                    employeeLinkedList.addLast(e);
//                }
//                startActivity();
                break;
//            case R.id.btn_loadFromFirebase:
//                new AsyncInitializerTask().execute(DataSource.FIREBASE);
//                break;
//            case R.id.btn_LoadFromSQLite:
//                new AsyncInitializerTask().execute(DataSource.SQLIT_DB);
//                break;
//            case R.id.btn_LoadFromWebAPI:
//                new AsyncInitializerTask().execute(DataSource.WEB_API);
//                break;
        }
    }

    public interface IsDataFilled {
        public void fillEmpList(boolean result);
    }

    private class AsyncInitializerTask extends AsyncTask<String, Void, Void> {

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            startActivity(new Intent(HR_MainActivity.this, EmployeeListActivity.class));
        }

        @Override
        protected Void doInBackground(String... strings) {
            fillEmpList(strings[0]);
            return null;
        }

        private void fillEmpList(String source) {
            switch (source) {
                case DataSource.INITIALIZE:
                    employeeLinkedList.clear();
                    Employee e;
                    int empID = 1;
                    String[] depts = {"Accounting", "HR", "IT"};
                    int[] avatars = {R.drawable.emp_pic1, R.drawable.emp_pic2, R.drawable.emp_pic3, R.drawable.emp_pic4, R.drawable.emp_pic5, R.drawable.emp_pic6, R.drawable.emp_pic7};
                    for (int i = 0; i < 20; i++) {
                        e = new Employee("Emplyee " + empID + " Name ", "Emplyee " + empID + " Email ", "Emplyee " + empID + " Phone ",
                                "Emplyee " + empID + " Address ", depts[i % 3], empID++, avatars[i % 7], Utils.getDateFromString("22.05.2016"));
                        employeeLinkedList.addLast(e);
                    }
                    break;
//                case DataSource.FIREBASE:
//                    firebaseDB_helper.getEmpRef().addListenerForSingleValueEvent(new ValueEventListener() {
//                        @Override
//                        public void onDataChange(DataSnapshot dataSnapshot) {
//                            if (dataSnapshot.hasChildren()) {
//                                employeeLinkedList.clear();
//                                Employee e;
//                                for (DataSnapshot ds : dataSnapshot.getChildren()) {
//                                    //Utils.getDateFromString("22.05.2016")
//                                    e = new Employee(ds.child("name").getValue(String.class)
//                                            , ds.child("email").getValue(String.class),
//                                            ds.child("phone").getValue(String.class),
//                                            ds.child("address").getValue(String.class),
//                                            ds.child("dept").getValue(String.class),
//                                            ds.child("id").getValue(Integer.class),
//                                            ds.child("picResID").getValue(Integer.class),
//                                            ds.child("hireDate").getValue(Date.class)
//                                    );
//                                    employeeLinkedList.addLast(e);
//                                }
//                                Intent intent = new Intent(MainActivity.this, EmployeeListActivity.class);
//                                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
//                                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
//                                startActivity(intent);
//                            } else {
//                                fillEmpList(DataSource.INITIALIZE);
//                                fillEmpList(DataSource.FIREBASE);
//                            }
//                        }
//
//                        @Override
//                        public void onCancelled(DatabaseError databaseError) {
//                            // ...
//                        }
//                    });
//                    break;
//                case DataSource.SQLIT_DB:
//                    SQLiteDB_Helper ss = SQLiteDB_Helper.getInstance(MainActivity.this);
//                    if (ss.getEmployeesCount() > 0) {
//                        employeeLinkedList.clear();
//                        employeeLinkedList = ss.getEmployees();
//                    } else {
//                        fillEmpList(DataSource.INITIALIZE);
//                        //ss.insertEmployees(employeeLinkedList);
//                        for (Employee ee : employeeLinkedList) {
//                            ss.insertEmployee(ee);
//                        }
//                        fillEmpList(DataSource.SQLIT_DB);
//                    }
//                    break;
//                case DataSource.WEB_API:
//                    new JsonTask(MainActivity.this).execute("http://192.168.137.1:8070/APIs/EmpAPIService.svc/getEmployees");
//
//                    break;
            }

        }

    }

    public class JsonTask extends AsyncTask<String, String, String> {

        private Context context;

        public JsonTask(Context context) {
            this.context = context;
        }

        protected void onPreExecute() {
            super.onPreExecute();
        }

        protected String doInBackground(String... params) {


            HttpURLConnection connection = null;
            BufferedReader reader = null;

            try {
                URL url = new URL(params[0]);
                connection = (HttpURLConnection) url.openConnection();
                connection.connect();


                InputStream stream = connection.getInputStream();

                reader = new BufferedReader(new InputStreamReader(stream));

                StringBuffer buffer = new StringBuffer();
                String line = "";

                while ((line = reader.readLine()) != null) {
                    buffer.append(line + "\n");
                }

                return buffer.toString();


            } catch (MalformedURLException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                if (connection != null) {
                    connection.disconnect();
                }
                try {
                    if (reader != null) {
                        reader.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return null;
        }

        @Override
        protected void onPostExecute(String result) {
            super.onPostExecute(result);
            Log.i("Result", result);
            try {
                JSONObject jsonObject = new JSONObject(result);
                String isSuccess = jsonObject.getString("status");
                if (isSuccess.equals("Success")) {
                    JSONArray empList = jsonObject.getJSONArray("data");
                    JSONObject emp = null;
                    Employee e = null;
                    HR_MainActivity.employeeLinkedList.clear();
                    for (int i = 0; i < empList.length(); i++) {
                        emp = empList.getJSONObject(i);
                        e = new Employee(emp.getString("name"),
                                emp.getString("email"),
                                emp.getString("phone"),
                                emp.getString("address"),
                                emp.getString("dept"),
                                emp.getInt("ID"),
                                emp.getInt("picResID"),
                                null);
                        HR_MainActivity.employeeLinkedList.add(e);
                    }
                    Intent intent = new Intent(getApplicationContext(), EmployeeListActivity.class);
                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/com/examples/android/androidtrainingbzu/HR_MainActivity.kt
package com.examples.android.androidtrainingbzu

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.examples.android.androidtrainingbzu.Database.FirebaseDB_Helper
import com.examples.android.androidtrainingbzu.Database.SQLiteDB_Helper
import com.examples.android.androidtrainingbzu.Models.Employee
import com.examples.android.androidtrainingbzu.Utils.DataSource
import com.examples.android.androidtrainingbzu.Utils.Utils
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.*
import java.util.concurrent.Executors

class HR_MainActivity : AppCompatActivity() {
    private val TAG = HR_MainActivity::class.java.simpleName
    var firebaseDB_helper: FirebaseDB_Helper? = null

    private val executor = Executors.newSingleThreadExecutor()
    private val handler = Handler(Looper.getMainLooper())

    companion object {
        var employeeLinkedList: LinkedList<Employee>? = LinkedList<Employee>()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_hr__main)
        firebaseDB_helper = FirebaseDB_Helper.Companion.getInstance(this)
    }

    fun startActivity() {
        val intent = Intent(applicationContext, EmployeeListActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        startActivity(intent)
    }

    fun callInitialize(view: View) {
        executor.execute {
            when (view.id) {
                R.id.btn_inialize -> {
                    fillListItemsFrom(DataSource.INITIALIZE)
                }
                R.id.btn_LoadFromSQLite -> {
                    fillListItemsFrom(DataSource.SQLIT_DB)
                }
                R.id.btn_LoadFromWebAPI -> {
                    fillListItemsFrom(DataSource.WEB_API)
                }
                R.id.btn_loadFromFirebase -> {
                    fillListItemsFrom(DataSource.FIREBASE)
                }
            }
        }
        handler.post {
            startActivity(Intent(this@HR_MainActivity, EmployeeListActivity::class.java))
        }
    }

    interface IsDataFilled {
        fun fillEmpList(result: Boolean)
    }

    private fun fillListItemsFrom(source: String?) {
        when (source) {
            DataSource.INITIALIZE -> {
                employeeLinkedList!!.clear()
                var e: Employee
                var empID = 1
                val depts = arrayOf("Accounting", "HR", "IT")
                val avatars = intArrayOf(R.drawable.emp_pic1,
                    R.drawable.emp_pic2,
                    R.drawable.emp_pic3,
                    R.drawable.emp_pic4,
                    R.drawable.emp_pic5,
                    R.drawable.emp_pic6,
                    R.drawable.emp_pic7)
                var i = 0
                while (i < 20) {
                    e = Employee("Emplyee $empID Name ",
                        "Emplyee $empID Email ",
                        "Emplyee $empID Phone ",
                        "Emplyee $empID Address ",
                        depts[i % 3],
                        empID++,
                        avatars[i % 7],
                        Utils.getDateFromString("22.05.2016"))
                    employeeLinkedList!!.addLast(e)
                    i++
                }
            }
            DataSource.SQLIT_DB -> {
                val sqlitedbHelper: SQLiteDB_Helper? =
                    SQLiteDB_Helper.Companion.getInstance(this@HR_MainActivity)

                if (sqlitedbHelper?.employeesCount!! > 0) {
                    employeeLinkedList!!.clear()
                    employeeLinkedList = sqlitedbHelper.employees
                } else {
                    sqlitedbHelper.insertEmployees(employeeLinkedList)
                    //                        for (Employee ee : employeeLinkedList) {
//                            ss.insertEmployee(ee);
//                        }
                    fillListItemsFrom(DataSource.SQLIT_DB)
                }
            }
            DataSource.WEB_API -> {
                var connection: HttpURLConnection? = null
                var reader: BufferedReader? = null
                try {
                    val url = URL("http://192.168.43.36:8070/APIs/EmpAPIService.svc/getEmployees")
//                    connection = url.openConnection() as HttpURLConnection
//                    connection.connect()
//                    val stream = connection.inputStream
//                    reader = BufferedReader(InputStreamReader(stream))
//                    val buffer = StringBuffer()
//                    var line = ""
//                    while (reader.readLine().also { line = it } != null) {
//                        buffer.append("""$line""".trimIndent())
//                    }

                    val jsonObject = JSONObject(url.readText())
                    val isSuccess = jsonObject.getString("status")
                    if (isSuccess == "Success") {
                        val empList = jsonObject.getJSONArray("data")
                        var emp: JSONObject? = null
                        var e: Employee? = null
                        employeeLinkedList!!.clear()
                        for (i in 0 until empList.length()) {
                            emp = empList.getJSONObject(i)
                            e = Employee(emp.getString("name"),
                                emp.getString("email"),
                                emp.getString("phone"),
                                emp.getString("address"),
                                emp.getString("dept"),
                                emp.getInt("ID"),
                                emp.getInt("picResID"),
                                null)
                            employeeLinkedList!!.add(e)
                        }
                    }
                } catch (e: MalformedURLException) {
                    e.printStackTrace()
                } catch (e: IOException) {
                    e.printStackTrace()
                } finally {
                    connection?.disconnect()
//                    try {
//                        reader!!.close()
//                    } catch (e: IOException) {
//                        e.printStackTrace()
//                    }
                }
            }

            DataSource.FIREBASE -> firebaseDB_helper?.empRef?.addListenerForSingleValueEvent(object :
                ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.hasChildren()) {
                        employeeLinkedList!!.clear()
                        var e: Employee
                        for (ds in dataSnapshot.children) {
                            //Utils.getDateFromString("22.05.2016")
                            e = Employee(ds.child("name").getValue(String::class.java),
                                ds.child("email").getValue(String::class.java),
                                ds.child("phone").getValue(String::class.java),
                                ds.child("address").getValue(String::class.java),
                                ds.child("dept").getValue(String::class.java),
                                ds.child("id").getValue(Int::class.java)!!,
                                ds.child("picResID").getValue(Int::class.java)!!,
                                ds.child("hireDate").getValue(Date::class.java)
                            )
                            employeeLinkedList!!.addLast(e)
                        }
                        val intent =
                            Intent(this@HR_MainActivity, EmployeeListActivity::class.java)
                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        startActivity(intent)
                    } else if (employeeLinkedList!!.size > 0) {
                        val empRef = firebaseDB_helper?.empRef
                        for (e in employeeLinkedList!!) {
                            empRef!!.child("Emp_" + e.id).setValue(e)
                        }
                        fillListItemsFrom(DataSource.FIREBASE)
                    } else {
                        fillListItemsFrom(DataSource.INITIALIZE)
                        fillListItemsFrom(DataSource.FIREBASE)
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    // ...
                }
            })
        }
    }
//    inner class AsyncInitializerTask : AsyncTask<String?, Void?, Void?>() {
//        override fun onPostExecute(aVoid: Void?) {
//            super.onPostExecute(aVoid)
//            startActivity(Intent(this@HR_MainActivity, EmployeeListActivity::class.java))
//        }
//
//        protected override fun doInBackground(vararg strings: String): Void? {
//            fillEmpList(strings[0])
//            return null
//        }
//
//        private fun fillEmpList(source: String?) {
//            when (source) {
////                DataSource.INITIALIZE -> {
////                    employeeLinkedList!!.clear()
////                    var e: Employee
////                    var empID = 1
////                    val depts = arrayOf("Accounting", "HR", "IT")
////                    val avatars = intArrayOf(R.drawable.emp_pic1,
////                        R.drawable.emp_pic2,
////                        R.drawable.emp_pic3,
////                        R.drawable.emp_pic4,
////                        R.drawable.emp_pic5,
////                        R.drawable.emp_pic6,
////                        R.drawable.emp_pic7)
////                    var i = 0
////                    while (i < 20) {
////                        e = Employee("Emplyee $empID Name ",
////                            "Emplyee $empID Email ",
////                            "Emplyee $empID Phone ",
////                            "Emplyee $empID Address ",
////                            depts[i % 3],
////                            empID++,
////                            avatars[i % 7],
////                            Utils.getDateFromString("22.05.2016"))
////                        employeeLinkedList!!.addLast(e)
////                        i++
////                    }
////                }
//                DataSource.FIREBASE -> firebaseDB_helper.getEmpRef()
//                    .addListenerForSingleValueEvent(object : ValueEventListener {
//                        override fun onDataChange(dataSnapshot: DataSnapshot) {
//                            if (dataSnapshot.hasChildren()) {
//                                employeeLinkedList!!.clear()
//                                var e: Employee
//                                for (ds in dataSnapshot.children) {
//                                    //Utils.getDateFromString("22.05.2016")
//                                    e = Employee(ds.child("name").getValue(String::class.java),
//                                        ds.child("email").getValue(String::class.java),
//                                        ds.child("phone").getValue(String::class.java),
//                                        ds.child("address").getValue(String::class.java),
//                                        ds.child("dept").getValue(String::class.java),
//                                        ds.child("id").getValue(Int::class.java)!!,
//                                        ds.child("picResID").getValue(Int::class.java)!!,
//                                        ds.child("hireDate").getValue(Date::class.java)
//                                    )
//                                    employeeLinkedList!!.addLast(e)
//                                }
//                                val intent =
//                                    Intent(this@HR_MainActivity, EmployeeListActivity::class.java)
//                                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
//                                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
//                                startActivity(intent)
//                            } else if (employeeLinkedList!!.size > 0) {
//                                val empRef = firebaseDB_helper.getEmpRef()
//                                for (e in employeeLinkedList!!) {
//                                    empRef!!.child("Emp_" + e.getID()).setValue(e)
//                                }
//                                fillEmpList(DataSource.FIREBASE)
//                            } else {
//                                fillEmpList(DataSource.INITIALIZE)
//                                fillEmpList(DataSource.FIREBASE)
//                            }
//                        }
//
//                        override fun onCancelled(databaseError: DatabaseError) {
//                            // ...
//                        }
//                    })
//                DataSource.SQLIT_DB -> {
//                    val ss: SQLiteDB_Helper =
//                        SQLiteDB_Helper.Companion.getInstance(this@HR_MainActivity)
//                    if (ss.employeesCount > 0) {
//                        employeeLinkedList!!.clear()
//                        employeeLinkedList = ss.employees
//                    } else {
//                        fillEmpList(DataSource.INITIALIZE)
//                        ss.insertEmployees(employeeLinkedList)
//                        //                        for (Employee ee : employeeLinkedList) {
////                            ss.insertEmployee(ee);
////                        }
//                        fillEmpList(DataSource.SQLIT_DB)
//                    }
//                }
//                DataSource.WEB_API -> JsonTask(this@HR_MainActivity).execute("http://172.19.16.180:8070/APIs/EmpAPIService.svc/getEmployees")
//            }
//        }
//    }
//
//    inner class JsonTask(private val context: Context) : AsyncTask<String?, String?, String?>() {
//        override fun onPreExecute() {
//            super.onPreExecute()
//        }
//
//        protected override fun doInBackground(vararg params: String): String? {
//            var connection: HttpURLConnection? = null
//            var reader: BufferedReader? = null
//            try {
//                val url = URL(params[0])
//                connection = url.openConnection() as HttpURLConnection
//                connection!!.connect()
//                val stream = connection.inputStream
//                reader = BufferedReader(InputStreamReader(stream))
//                val buffer = StringBuffer()
//                var line = ""
//                while (reader.readLine().also { line = it } != null) {
//                    buffer.append("""
//    $line
//
//    """.trimIndent())
//                }
//                return buffer.toString()
//            } catch (e: MalformedURLException) {
//                e.printStackTrace()
//            } catch (e: IOException) {
//                e.printStackTrace()
//            } finally {
//                connection?.disconnect()
//                try {
//                    reader?.close()
//                } catch (e: IOException) {
//                    e.printStackTrace()
//                }
//            }
//            return null
//        }
//
//        override fun onPostExecute(result: String?) {
//            super.onPostExecute(result)
//            Log.i("Result", result!!)
//            try {
//                val jsonObject = JSONObject(result)
//                val isSuccess = jsonObject.getString("status")
//                if (isSuccess == "Success") {
//                    val empList = jsonObject.getJSONArray("data")
//                    var emp: JSONObject? = null
//                    var e: Employee? = null
//                    employeeLinkedList!!.clear()
//                    for (i in 0 until empList.length()) {
//                        emp = empList.getJSONObject(i)
//                        e = Employee(emp.getString("name"),
//                            emp.getString("email"),
//                            emp.getString("phone"),
//                            emp.getString("address"),
//                            emp.getString("dept"),
//                            emp.getInt("ID"),
//                            emp.getInt("picResID"),
//                            null)
//                        employeeLinkedList!!.add(e)
//                    }
//                    val intent = Intent(applicationContext, EmployeeListActivity::class.java)
//                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
//                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
//                    startActivity(intent)
//                }
//            } catch (e: JSONException) {
//                e.printStackTrace()
//            }
//        }
//    }


}